src/app/rundown-view/services/mini-shelf-state.service.ts

```diff
@@ -0,0 +1,114 @@
+import { Injectable } from '@angular/core'
+import { Rundown } from '../../core/models/rundown'
+import { Segment } from '../../core/models/segment'
+import { ActionService } from '../../shared/abstractions/action.service'
+import { RundownEventObserver } from '../../core/services/rundown-event-observer.service'
+import { PartTakenEvent, RundownActivatedEvent } from '../../core/models/rundown-event'
+import { Tv2VideoClipAction } from '../../shared/models/tv2-action'
+
+@Injectable()
+export class MiniShelfStateService {
```

Comment:
In the project, we have a concept of state services which are classes that are responsible for holding the state of the subject they describe. The state services have internal mechanisms that react to events from the backend (or another source) in order to update and notify any subscribers about changes to the state.

Since this service seems to be told what the state is in order to compute and act on the given data, and does not expose the state for other classes to depend on, I don't see it as a state service in the same sense.

I suggest that the different responsibilities be separated out into their own classes, such that there is one responsibility of navigating/finding mini shelf groups; one that is handling the mini shelf state; and a mediator, which is allowed to use the two others to coordinate and act upon the outcome. By having the navigation and searching for mini shelf groups in a separate class, we have isolated the business logic of the feature and can easily test it.

 I have added a quickly sketched diagram below as an example. Note that this is just a proposal and should at best be used as inspiration :)

```mermaid
graph TD

RundownStateService["
  Runndown State Service
  [Service]
"]

ActionStateService["
  Action State Service
  [Service]
"]

MediaStateService["
  Media State Service
  [Service]
"]

MiniShelfStateService["
  Mini Shelf State Service
  [Service]

  Is responsible for merging action and media data
  with mini shelf segments and expose them via
  observables.
"]

MiniShelfNavigationService["
  Mini Shelf Navigation Service
  [Service]

  Given a rundown and the current mini
  shelf segment, sit calculates what the next
  should be.
"]

MiniShelfCycleService["
  Mini Shelf Cycle Service
  [Service]

  Is the mediator that can be called to orchestrate
  the cycling and the execution of it.
"]

ActionService["
  Action Service
  [Service]

  Is responsible for executing actions.
"]

KeyBindingService["
  Key Binding Service
  [Service]
"]

KeyBindingService --> |"triggers minishelf cycling via"| MiniShelfCycleService

MiniShelfCycleService --> |"asks for next/prev mini shelf from"| MiniShelfNavigationService
MiniShelfCycleService --> |"subscribes to mini shelf data from"| MiniShelfStateService
MiniShelfCycleService --> |"executes the mini shelf action via"| ActionService

MiniShelfStateService --> |"subscribes to actions from"| ActionStateService
MiniShelfStateService --> |"subscribes to rundowns from"| RundownStateService
MiniShelfStateService --> |"subscribes to media from"| MediaStateService
```